linked_list/src/lib.rs

```diff
@@ -0,0 +1,210 @@
+use std::cell::RefCell;
+use std::rc::Rc;
+
+// Тип, который представляет разделяемый и изменяемый указатель на узел
+type Link<T> = Option<Rc<RefCell<Node<T>>>>;
+
+// Узел списка
+#[derive(Debug)]
+pub struct Node<T> {
+    pub value: T,
+    pub next: Link<T>,
+}
+
+// Связанный список
+#[derive(Debug)]
+pub struct LinkedList<T> {
+    pub head: Link<T>,
+    tail: Link<T>,
+}
+
+impl<T> LinkedList<T> {
+    fn new() -> Self {
+        LinkedList {
+            head: None,
+            tail: None,
+        }
+    }
+
+    // Добавление элемента в начало
+    pub fn push_front(&mut self, value: T) {
+        let new_node = Rc::new(RefCell::new(Node {
+            value,
+            next: self.head.clone(),
+        }));
+
+        self.head = Some(new_node.clone());
+        if self.tail.is_none() {
+            self.tail = Some(new_node);
+        }
+    }
+
+    // Добавление элемента в конец
+    pub fn push_back(&mut self, value: T) {
+        let new_node = Rc::new(RefCell::new(Node { value, next: None }));
+
+        match self.tail.take() {
+            Some(old_tail) => {
+                old_tail.borrow_mut().next = Some(new_node.clone());
+            }
+            None => {
+                self.head = Some(new_node.clone());
+            }
+        }
+        self.tail = Some(new_node);
+    }
+
+    // Добавление элемента после N-го
+    pub fn insert_after(&mut self, index: usize, value: T) {
+        let mut current = self.head.clone();
+        for _ in 0..index {
+            match current {
+                Some(ref node) => {
+                    let next = node.borrow().next.clone();
+                    current = next;
+                }
+                None => return,
+            }
+        }
+
+        if let Some(ref node) = current {
+            let new_node = Rc::new(RefCell::new(Node {
+                value,
+                next: node.borrow().next.clone(),
+            }));
+            let mut node_mut = node.borrow_mut();
+            if node_mut.next.is_none() {
+                self.tail = Some(new_node.clone());
+            }
+            node_mut.next = Some(new_node);
+        }
+    }
+
+    // Итератор по всем элементам
+    pub fn iter(&self) -> LinkedListIterator<T> {
+        LinkedListIterator {
+            current: self.head.clone(),
+        }
+    }
+
+    // Разделение списка на два списка
+    pub fn split_at(&mut self, index: usize) -> LinkedList<T> {
```

Comment:
Предлагаю тут переписать вот так:

```suggestion
    pub fn split_at(self, index: usize) -> (Self, Self) {
```

Типа это разделение одного списка на два, так что тут такое вполне логичным кажется.